lib/block-supports/layout.php

```diff
@@ -162,6 +162,17 @@ function gutenberg_render_layout_support_flag( $block_content, $block ) {
 
 	$class_name = wp_unique_id( 'wp-container-' );
 	$gap_value  = _wp_array_get( $block, array( 'attrs', 'style', 'spacing', 'blockGap' ) );
+
+	// If there is no block-level value for blockGap,
+	// but a global styles value available for blockGap,
+	// use the latter.
+	if ( empty( $gap_value ) ) {
```

Comment:
Why is this needed? isn't the fallback supposed to happen using the CSS variable?

Reply:
(And we can only set the CSS variable at the global level theme.json)

Reply:
> Why is this needed?

Good question. 

Logically it's not. 

I only put it there so we could skip over trying to fetch the global value via `gutenberg_get_global_styles` if a gap value set in the block's local styles attribute is found, which would take precedence anyway.

> isn't the fallback supposed to happen using the CSS variable? 

It will either way, unless I'm missing something, which could be the case 😇 

My thinking went along of these lines: if the block's local styles attribute does not contain a gap value, we'll try the block's global styles. 

Either could still be `null` by the time we call `gutenberg_get_layout_style()` in which case the fallback would be `--wp--style--block-gap`, which is set at the root. 

Does that sound right?

Reply:
It does feel like duplication to me, since if the first one is null, the fallback will be the CSS variable (which corresponds to the root value). 

Or in other words, if we do the fallback in php, why do we have the CSS variable in the first place.

Reply:
Oh really sorry @ramonjd I misunderstand the original purpose of this PR :) 

Now that I understand it better, I think it's actually a smart way to solve "gap" support in theme.json at all levels. 

That said, I'm still hesitant for a couple reasons:

 - This basically changes how the "style engine" of layout works, instead of being `styleObject => CSS`, it's now `( styleObject, theme.json ) => CSS`. It changes the mental model and will have an impact on the style engine work.
 - In #39281 we're planning to add "theme.json" like support to each block (containers), meaning a container a "group" block will be able to say: all social link blocks inside me should have a gap of 10px. Let's think a little bit about what that means. It means that we can have nested containers which each container defining a different block gap (or not) until we reach the root level. It means that to generate the right style here we need to be aware of all the parent blocks' theme.json. So instead of `( styleObject, theme.json ) => CSS`, it becomes `( styleObject, theme.jsonForAllMyContainers, rootTheme.json) => CSS`. I think it's probably very hard (if not impossible) to access the `theme.jsonForAllMyContainers` from a given block.

Reply:
You raise good points, thanks for clarifying. 

> This basically changes how the "style engine" of layout works, instead of being styleObject => CSS, it's now ( styleObject, theme.json ) => CSS. It changes the mental model and will have an impact on the style engine work.

Yeah, I see what you mean. We might run into trouble if we mix block support style attributes with global styles. 

> we're planning to add "theme.json" like support to each block (containers)

I wasn't aware of Style Partials (theme.json) so I agree that a change like the one in this PR might come back to bite us. 👍 

I was having a discussion with @tellthemachines about something similar over at [Try using variable for root padding value #39926](https://github.com/WordPress/gutenberg/pull/39926#discussion_r839253862)

I was wondering about an alternative: a `const VALID_TOP_LEVEL_STYLES` (or whatever) in `WP_Theme_JSON_Gutenberg` that explicitly defines unique, root-level-only CSS vars.

```
body {
    --wp--root--style--block-gap: 2px;
    --wp--root--style--padding: 10px;
     /* etc ... */
}
```

I haven't really thought this through, but I was venturing towards checking `VALID_TOP_LEVEL_STYLES` when we assemble the styles in `compute_style_properties` and appending `--wp--root-style--` for  `ROOT_BLOCK_SELECTOR` styles.

Or there might be a tricky way to assemble them in `sanitize()` via `$schema['styles']`, `$schema['styles']['elements']` and `$schema['styles']['blocks']`. 

Uniquely identifying global CSS vars has a benefit, aside from explicitly stating their purpose, in that their values can be used for top-level layout calculation. 

For example, in https://github.com/WordPress/gutenberg/pull/39926, full-width blocks can use a root padding CSS var set a negative margins so that they stretch to the "full-width" of the screen.

Ultimately, we need to know how to treat `blockGap` when it is a root or a block/element style value and output something different accordingly. 

I can chat about it with @andrewserong and see what we can cook up.

Reply:
Great questions.

> This basically changes how the "style engine" of layout works, instead of being styleObject => CSS, it's now ( styleObject, theme.json ) => CSS. It changes the mental model and will have an impact on the style engine work.

Totally, I think it reveals some pragmatic challenges surrounding how we output styles, particularly for the layout support, which as @ramonjd mentions further down, is a little different to the other block supports.

In the case of this PR (and the current implementation of the layout support), everything happens at the individual block level when it's rendered, so out of necessity, to factor in global styles, we've kinda got to "look up" to the parents in order to deal with those values (or use CSS variables to stand-in for it).

In an ideal circumstance, with the style engine, how differently might we be doing things? The model for the style engine is a lot clearer for supports like padding, margin, typography, etc, where we'd like to consolidate the approach at the individual block level and global styles, and using a block's classname is an appropriate hook for it. In some of the explorations (like https://github.com/WordPress/gutenberg/pull/39374) it looks like we'll be able to move more to global styles, but the unique block gap values was still a bit of mystery to me as to how best to handle it, and neatly declare values that get passed down / or cascade down correctly.

With block gap, because the value is used directly in different circumstances (rendering `flow` vs `flex` gap), it does create this additional challenge. I think because the use case is so different to each of the other block supports, I'm wondering if the approach in this PR would be an acceptable way to unblock getting the global styles support in, and if we could treat it as an implementation detail we could potentially swap out if and when partials in https://github.com/WordPress/gutenberg/issues/39281 is explored.

From my perspective, it could be worthwhile proceeding with this change, so long as:

* We're happy that folks will set `blockGap` values at the block level within `theme.json` and via the global styles interface in the site editor.
* So long as we're happy with the shape of the data as it's stored, then theoretically, we could change the implementation after the fact, once we have a clearer idea of how we'd like to handle outputting the layout styles in the style engine.

From some of the explorations we've looked at recently, given that a fair bit of the logic winds up being moved to the Theme JSON class anyway, I don't _think_ that the current change here would preclude further work, but very happy to be proved wrong, as I definitely don't want to nudge us into a decision that's difficult to reverse!

Reply:
> With block gap, because the value is used directly in different circumstances (rendering flow vs flex gap), it does create this additional challenge. 

One alternative could be to keep the current logic (selectors specific styles but output something like):

```
.selector.is-layout-flex {
   gap: something;
}

.selector.is-layout-default > * {
   margin-top: something;
}
```

To be honest, I didn't think much about this, it's just a random idea for now, but it's an alternative that allows us to avoid that dependency to parent global style.

 > From some of the explorations we've looked at recently, given that a fair bit of the logic winds up being moved to the Theme JSON class anyway, I don't think that the current change here would preclude further work, but very happy to be proved wrong

I don't know 🤷 to be honest, personally I don't have a full confidence here but I trust your judgement.

Reply:
> One alternative could be to keep the current logic (selectors specific styles but output something like):

I really like that approach — I'd like it if we can output something like that in the longer term. I might be overthinking it, but I think that approach slightly opens up the can-of-worms surrounding rendering out semantic classnames, which we're in more of the proposal / exploratory phase of (since we'd need to adjust the logic for how the theme JSON class outputs styles, and then the layout support to inject the classnames).

> I don't know 🤷 to be honest, personally I don't have a full confidence here but I trust your judgement.

Thank you, much appreciated! Since I was one of the folks to propose the idea of looking up global styles, I think I'd feel more comfortable with getting a couple more opinions if you're not totally sold on the idea, to confidence check before proceeding. I think my main interest is to see how viable we think a short-term fix is (so that we can fix this bug in WP 6.0) and then aim to have the class-based approach like in your code snippet in time for WP 6.1.